Hometask 4

kkonanov opened the pull request.

Commit:
[task2]
add interface Actionable
add functional interface Action
release inner iterator by method execute(Action action)
add unit-tests for DoorEventHandler, LightEventHandler, AllLightEventHandler, SmartHome.execute, Room.execute, Light.execute, Door.execute

Commit:
[task3]
add Signalization and State

Commit:
[task3]
add SignalizationEventHandler and Signalization

kkonanov committed:
[task3]

kkonanov committed:
[task4]
add Spring realization and using API

kkonanov committed:
[task4]

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHomeConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	@Bean
+	Collection<EventHandler> eventHandlers() {
```
Не надо создавать коллекцию, это спринг сделает за вас.
Объявляйте отдлеьными бинами обарботчики, добавляйте в каждый декоратор, чтобы клиент не знал про декоратор.
https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BE%D0%B1%D1%8A%D1%8F%D0%B2%D0%BB%D1%8F%D0%B9%D1%82%D0%B5-%D0%BA%D0%B0%D0%B6%D0%B4%D0%BE%D0%B3%D0%BE-%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%87%D0%B8%D0%BA%D0%B0-%D0%BE%D1%82%D0%B4%D0%B5%D0%BB%D1%8C%D0%BD%D1%8B%D0%BC-%D0%B1%D0%B8%D0%BD%D0%BE%D0%BC

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHomeConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	@Bean
+	SensorEventsManager sensorEventsManager() {
```
`@Bean
SensorEventsManager sensorEventsManager(Collection<EventProcessor> eventProcessors) {...}`

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SensorEventsManagerAdapter.java:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	private EventType getEventTypeByString(String CSSEventType) {
+		switch (CSSEventType) {
```
Нарушение OCP, см. https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%B8%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D1%83%D0%B9%D1%82%D0%B5-%D1%84%D0%B0%D0%B1%D1%80%D0%B8%D0%BA%D0%B8-%D0%B4%D0%BB%D1%8F-%D0%BA%D0%BE%D0%BD%D0%B2%D0%B5%D1%80%D1%82%D0%B0%D1%86%D0%B8%D0%B8-%D1%81%D0%BE%D0%B1%D1%8B%D1%82%D0%B8%D0%B9-%D0%B2%D0%BD%D0%B5%D1%88%D0%BD%D0%B5%D0%B9-%D0%B1%D0%B8%D0%B1%D0%B8%D0%BB%D0%BE%D1%82%D0%B5%D0%BA%D0%B8

kkonanov committed:
delete @Bean eventHandlers
create ConversionFactory
change SensorEventsManagerAdapter

kkonanov commented on src/main/java/ru/sbt/mipt/oop/SmartHomeConfiguration.java:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	@Bean
+	SensorEventsManager sensorEventsManager(Collection<EventHandler> eventHandlers, ConversionFactory conversionFactory) {
```
Как сделать так, чтобы коллекция составлялась в определенном порядке (с @Order не получилось, см. выше)

a-ivanov replied:
Использовать соответствующий интерфейс с порядком, например, List.
В Collection можно просто добавить объект, итерироваться по ней в каком-то порядке.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SensorEventsManagerAdapter.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+
+@Component
+public class SensorEventsManagerAdapter {
```
Это не адаптер, это, скорее, фабрика.
Адаптер должен имплементировать целевой интерфейс и делегировать обработку объектам из предыдущих заданий.
См. https://refactoring.guru/ru/design-patterns/adapter

kkonanov committed:
[hometask-4]
delete SensorEventsManagerAdapter.java
add EventHandlerAdapter (implements com.coolcompany.smarthome.events.EventHandler)
change configuration file

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/factory/ConversionFactory.java:
```diff
@@ -0,0 +1,27 @@
+	}
+
+	public SensorEvent createSensorEvent(CCSensorEvent ccSensorEvent) {
+		return new SensorEvent(stringToEventType.get(ccSensorEvent.getEventType()), ccSensorEvent.getObjectId());
```
Проверка на присутствие eventType в stringToEventType.

kakonanov commented:
Спасибо, исправлю)
